src/MizarData/include/MizarData/FrameTrack.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2022 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef MIZAR_DATA_MIZAR_FRAME_TRACK_H_
+#define MIZAR_DATA_MIZAR_FRAME_TRACK_H_
+
+#include "ClientData/ScopeId.h"
+#include "ClientData/ScopeInfo.h"
+#include "GrpcProtos/capture.pb.h"
+#include "OrbitBase/Typedef.h"
+#include "stdint.h"
+
+namespace orbit_mizar_data {
+
+struct FrameTrackIdTag {};
+// The type is used identify a frame track in Mizar. It can be either ETW or a scope.
+using FrameTrackId =
+    orbit_base::Typedef<FrameTrackIdTag, std::variant<orbit_client_data::ScopeId,
+                                                      orbit_grpc_protos::PresentEvent::Source>>;
+
+struct FrameStartNsTag {};
+// Absolute timestamp of the capture start in nanos
+using FrameStartNs = orbit_base::Typedef<FrameStartNsTag, uint64_t>;
+
+struct FrameTrackInfoTag {};
+// The class describes a frame track
+using FrameTrackInfo =
+    orbit_base::Typedef<FrameTrackInfoTag, std::variant<orbit_client_data::ScopeInfo,
+                                                        orbit_grpc_protos::PresentEvent::Source>>;
+
+// The function allows for more convenient calls to `std::visit`.
+// `Host` is a `Typedef` wrapping an `std::variant<First, Second>`.
+template <typename First, typename Second, typename Host, typename ActionOnFirst,
+          typename ActionOnSecond>
+auto Visit(ActionOnFirst&& action_on_scope_info, ActionOnSecond&& action_on_etw_source,
+           Host&& host) {
+  return std::visit(
+      [&action_on_scope_info, &action_on_etw_source](const auto& info) {
+        using InfoT = std::decay_t<decltype(info)>;
+        if constexpr (std::is_same_v<InfoT, First>) {
+          return std::invoke(action_on_scope_info, info);
+        }
+        if constexpr (std::is_same_v<InfoT, Second>) {
+          return std::invoke(action_on_etw_source, info);
+        }
+      },
+      *std::forward<Host>(host));
+}
+
+template <typename ActionOnScopeInfo, typename ActionOnEtwSource>
+auto Visit(ActionOnScopeInfo&& action_on_scope_info, ActionOnEtwSource&& action_on_etw_source,
+           const FrameTrackInfo& info) {
+  return Visit<orbit_client_data::ScopeInfo, orbit_grpc_protos::PresentEvent::Source>(
+      std::forward<ActionOnScopeInfo>(action_on_scope_info),
+      std::forward<ActionOnEtwSource>(action_on_etw_source), info);
+}
+
+template <typename ActionOnScopeId, typename ActionOnEtwSource>
+auto Visit(ActionOnScopeId&& action_on_scope_id, ActionOnEtwSource&& action_on_etw_source,
+           const FrameTrackId& id) {
+  return Visit<orbit_client_data::ScopeId, orbit_grpc_protos::PresentEvent::Source>(
+      std::forward<ActionOnScopeId>(action_on_scope_id),
+      std::forward<ActionOnEtwSource>(action_on_etw_source), id);
+}
```

Comment:
Otherwise, maybe consider the "overloaded" definition in https://en.cppreference.com/w/cpp/utility/variant/visit to make sure to get a compilation error if a type is not treated?

```
template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;

  std::visit(overloaded{[](ScopeId& scope_id) { /*...*/ },
                        [](PresentEvent::Source& source) { /*...*/ }}, *id);
```

Reply:
1. You're right, I want a compile-time check for exhaustiveness of the visitor. 
2. > This would also have the advantage of being trivially extensible if we ever wanted to add a FrameTrack type

     And my compile-time `if`s are just as extensible. And it's wrapped in a function, we will not have to extend at every call cite. 
3. IMO `overloaded ` is not any easier to understand. The code we have now, almost reads as plain-English. 
4. `overloaded` works only for functors, I wish to also pass function pointers.

Reply:
>  The code we have now, almost reads as plain-English.

Surely you are joking :-)

> And my compile-time ifs are just as extensible. And they're wrapped in a function, we will not have to extend at every call cite.

If one wants to add types in the `Host` std::variant, they do have to modify the call site to handle the new type. They also need to change the templated dispatch code. My preference would be to have everything at the call site, something like:

```
  [[nodiscard]] std::vector<FrameStartNs> GetFrameStarts(FrameTrackId id, FrameStartNs min_start,
                                                         FrameStartNs max_start) const {
    return std::visit(
        [this, min_start, max_start](auto&& typed_id) {
          using T = std::decay_t<decltype(typed_id)>;
          if constexpr (std::is_same_v<T, ScopeId>)
            return GetScopeFrameStarts(min_start, max_start, typed_id);
          else if constexpr (std::is_same_v<T, PresentEvent::Source>)
            return GetEtwFrameStarts(min_start, max_start, typed_id);
          else
            static_assert(always_false_v<T>, "non-exhaustive visitor!");
        },
        *id);
  }
```

instead of the current code, which needs matching boilerplate for every supported type:

```
  [[nodiscard]] std::vector<FrameStartNs> GetFrameStarts(FrameTrackId id, FrameStartNs min_start,
                                                         FrameStartNs max_start) const {
    return Visit(
        absl::bind_front(&FrameTrackManagerTmpl::GetScopeFrameStarts, this, min_start, max_start),
        absl::bind_front(&FrameTrackManagerTmpl::GetEtwFrameStarts, this, min_start, max_start),
        id);
  }
```

To me, what's happening is clearer and the code is simpler to maintain. If we add types to the std::variant, we get a "static assert" at the exact line where we need to modify code.

In the end, it's up to you, but I find it weird that the boilerplate templated code needs to know about the number of number of types and the actual types. Any thoughts @beckerhe ?

Reply:
>  I find it weird that the boilerplate templated code needs to know about the number of number of types and the actual types

Actually, I like this point. I'll see how I can leverage type deduction.

Reply:
I've got rid of the two visit "specialisations".

Reply:
That's nice. However it still only supports 2 types. Adding a type means modifying the helper. Also, if we wanted to use the same pattern for a different number of types elsewhere, we'd have to duplicate the helper...

Reply:
True. I don't think this can be done without much more template magic. If a need to use `Visit` with different number of alternatives arises, I might reconsider in favour of `overloaded` from cppreference.